core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -164,6 +167,109 @@ private void subscribe(SubscriptionImpl subscription, Subscriber<Object> subscri
      */
     protected void subscribe0(EventExecutor executor, SubscriptionOption[] options) {}
 
+    @Override
+    public CompletableFuture<List<T>> collect(EventExecutor executor, SubscriptionOption... options) {
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        final CompletableFuture<List<T>> collectingFuture = new CompletableFuture<>();
+        final SubscriptionImpl subscription =
+                new SubscriptionImpl(this, NoopSubscriber.get(), executor, options, collectingFuture);
+        if (subscriptionUpdater.compareAndSet(this, null, subscription)) {
+            if (setState(State.CLOSED, State.CLEANUP)) {
+                final boolean withPooledObjects = subscription.withPooledObjects();
+                if (executor.inEventLoop()) {
+                    collectAll(collectingFuture, executor, withPooledObjects, true);
+                } else {
+                    executor.execute(() -> collectAll(collectingFuture, executor, withPooledObjects, false));
+                }
+            }
+        } else {
+            final Subscriber<Object> subscriber = this.subscription.subscriber();
+            final Throwable cause;
+            if (subscriber instanceof AbortingSubscriber) {
+                cause = ((AbortingSubscriber<Object>) subscriber).cause();
+            } else {
+                cause = new IllegalStateException("subscribed by other subscriber already");
+            }
+            collectingFuture.completeExceptionally(cause);
+        }
+        return collectingFuture;
+    }
+
+    private void collectAll(CompletableFuture<List<T>> collectingFuture, EventExecutor executor,
+                            boolean withPooledObjects, boolean directExecution) {
+        try {
+            collectingFuture.complete(drainAll(withPooledObjects, true));
+            // whenComplete() should be completed after executing the callbacks of collect().
+            if (directExecution) {
+                executor.execute(() -> whenComplete().complete(null));
+            } else {
+                whenComplete().complete(null);
+            }
```

Comment:
Actually, the original code is correct. My comment was not enough to explain the behavior. 😅
As you know, `collectAll` could be directly executed by the current thread, if the current thread is an event loop for subscribing.
Then we need to defer to complete `whenComplete()` so that the callbacks already attached to `collect()` is executed before completing `whenComplete()`.

However, the current thread is not the specified event loop, the completion of `collectingFuture` is executed asynchronously. That means we can not guarantee the execution order strongly. I thought an additional `executor.execute()` is meaningless for `whenComplete()`.

Reply:
> However, the current thread is not the specified event loop,

Isn't this method called by the specified event loop?
Which means we need to do `executor.execute(...)` no matter the value of `directExecution` if we want the future to be done after the callbacks of the collect future is notified?

Reply:
> Isn't this method called by the specified event loop?

What I meant was the caller thread.

>  Which means we need to do executor.execute(...) no matter the value of directExecution if we want the future to be done after the callbacks of the collect future is notified?

If the caller thread is the specified event loop, we are sure 100% that the synchronously attached callbacks will be always executed first and `executor.execute(() -> whenComplete().complete(null))` later. However, if the caller thread is a different thread, the `executor.execute(() -> whenComplete().complete(null))` could be completed first and the synchronously attached callbacks is executed later. So I only added `executor.execute(() -> ...)` for the indirection execution.

Because of these restrictions, I added some defense code to `FilteredStreamMessage` to adjust the mess-up order in case.

Reply:
Thanks for the explanation. 👍